src/Npgsql/NpgsqlCommand.cs

```diff
@@ -438,63 +442,172 @@ public new NpgsqlParameter CreateParameter()
+                        _connector.ReadExpecting<ParameterDescriptionMessage>();
+                        var msg = _connector.ReadMessage(DataRowLoadingMode.NonSequential);
+                        switch (msg.Code)
+                        {
+                        case BackendMessageCode.RowDescription:
+                            var description = (RowDescriptionMessage)msg;
+                            FixupRowDescription(description, isFirst);
+                            statement.Description = description;
+                            break;
+                        case BackendMessageCode.NoData:
+                            statement.Description = null;
+                            break;
+                        default:
+                            throw _connector.UnexpectedMessageReceived(msg.Code);
+                        }
+                        statement.IsPrepared = true;
+                        if (persist)
+                        {
+                            _connector.PersistentPreparedStatements.Add(statement.SQL, new PreparedStatementInfo
+                            {
+                                Name = statement.PreparedStatementName,
+                                Description = statement.Description
+                            });
+                        }
+                        else
+                            _connector.PreparedStatements.Add(statement.PreparedStatementName);
+                        isFirst = false;
                     }
+
+                    _connector.ReadExpecting<ReadyForQueryMessage>();
                 }
+
+                IsPrepared = true;
+                IsPersistent = persist;
+
+                Debug.Assert(Statements.All(s => s.IsPrepared));
+                Debug.Assert(Statements.All(s => s.PreparedStatementName != null));
             }
         }
 
-        void DeallocatePrepared()
+        void Unprepare()
         {
             if (!IsPrepared)
                 return;
+
+            // For persistent prepared commands, we don't actually Close the server-side statements,
+            // we just detach our statements from the server-side ones.
+            if (IsPersistent)
+            {
+                foreach (var statement in Statements)
+                    statement.Unprepare();
+                IsPrepared = false;
+                IsPersistent = false;
+            }
+            else
+            {
+                foreach (var statement in Statements)
+                    _connector.PreparedStatements.Remove(statement.PreparedStatementName);
+                ClosePreparedStatements();
+            }
+        }
+
+        /// <summary>
+        /// Unpersists a persistent prepared command.
+        /// </summary>
+        public void Unpersist()
+        {
+            // We want to allow users to unpersist without having to persist first
+            if (!IsPersistent)
+                Prepare(true);
```

Comment:
Yes, that's true, but it doesn't seem very important. Remember this is just Unpersist, which is supposed to be call once in the application lifecycle mostly. Also, it's within the programmer's ability (and responsibility) to avoid this scenario by making sure they call Unpersist() only for SQL that has been persisted before. Even if they don't, the consequences are minimal (prepare/unprepare)...